### PR TITLE
feat(gatsby-source-mongodb): accommodate relationships with MongoDB collections

### PR DESCRIPTION
**Pull request by wesleylhandy**
## Description

To accommodate relations between collections, I did the following. Firstly, I imported an instance of `ObjectID` from the `mongodb` package. Then, instead of calling `toString` on the _id, I used the function provided on the `ObjectID` instance `toHexString`. From there, for each key/value pair within an imported document, I created a recursive function that looks for nested instances of `ObjectID` and returned that value after converted `toHexString`.

My only concern involves performance on large datasets. I would appreciate any feedback concerning this.

## Related Issues

This fixes [issue 11587](https://github.com/gatsbyjs/gatsby/issues/11587) that identifies a bugs when importing documents that are relationships to other collections, such as fields that are another `ObjectID` or an array of `ObjectID`s.

I tested this using a [dataset](https://github.com/jonniebigodes/test_gatsby_source_mongo_relationships/blob/master/db_dump/) provided by @jonniebigodes to help debug this issue. Below is a screenshot indicating the preservation of the nested `ObjectID`s.

Screen Shot 2019-03-22 at 4 42 47 PM

### CODE REVIEW

**Comment by wesleylhandy** [comment]
Not sure why I'm failing, it passed until I pushed new code. But still, I want to refactor tests to make them more specific and make this an opt in feature by setting an optional parameter on configuring the plugin. I'll be able to get back to this on Monday.

**Comment by wesleylhandy** [comment]
Update:

I added more specific checks within testing, I also reconfigured this feature to be opt-in only via a new plugin option. Since MongoDB is a document model and not a relational model database, there will be many more collections that do not need this feature than those who do (maybe I'm making an assumption, but it can be turned on by setting `preserveObjectIds: true` within the `options` object.

**Comment by jonniebigodes** [comment]
@wesleylhandy if you don't mind i'll go through it with 2 eyes open/pardon the bad pun) today when i get home.

**Comment by wesleylhandy** [comment]
Thanks @jonniebigodes I fixed typo in README and then merged master into this branch, which seems to be what it takes to pass tests after making commits. I'm hopeful in a few minutes it will pass tests, which you will see next time you are online.

**Comment by jonniebigodes** [comment]
@wesleylhandy looks like all greens 👍 ! Also i've spent yesterday and today doing some tests to see how this change would behave. I've grabbed your code, created a free tier mongodb atlas cluster, added some data in there, started small and went to a larger dataset  and here are the results:

database: one_hundred_items
collections:
- categories 20 documents
- comments 100 documents
- posts 100 documents with 3 categories each
- users 20 documents

Gatsby output:
```bash
success open and validate gatsby-configs — 0.015 s
success load plugins — 0.229 s
success onPreInit — 0.383 s
success delete html and css files from previous builds — 0.012 s
success initialize cache — 0.015 s
success copy gatsby files — 0.479 s
success onPreBootstrap — 0.017 s
success source and transform nodes — 3.290 s
success building schema — 0.372 s
success createPages — 0.000 s
success createPagesStatefully — 0.089 s
success onPreExtractQueries — 0.000 s
success update schema — 0.179 s
success extract queries from components — 0.040 s
success run graphql queries — 0.028 s — 2/2 72.31 queries/second
success write out page data — 0.028 s
success write out redirect data — 0.002 s
success onPostBootstrap — 0.003 s

info bootstrap finished - 11.887 s

 DONE  Compiled successfully in 4535ms
```
database: one_thousand_items
collections:
- categories 50 documents
- comments 1000 documents
- posts 1000 documents with 3 categories each
- users 50 documents

Gatsby output:
```bash
success open and validate gatsby-configs — 0.011 s
success load plugins — 0.247 s
success onPreInit — 0.409 s
success delete html and css files from previous builds — 0.008 s
success initialize cache — 0.018 s
success copy gatsby files — 0.519 s
success onPreBootstrap — 0.022 s
success source and transform nodes — 5.114 s
success building schema — 0.445 s
success createPages — 0.033 s
success createPagesStatefully — 0.136 s
success onPreExtractQueries — 0.001 s
success update schema — 0.262 s
success extract queries from components — 0.040 s
success run graphql queries — 0.040 s — 2/2 51.19 queries/second
success write out page data — 0.033 s
success write out redirect data — 0.002 s
success onPostBootstrap — 0.002 s

info bootstrap finished - 14.465 s

 DONE  Compiled successfully in 4340ms
```

database: two_thousand_items
collections:
- categories 100 documents
- comments 2000 documents
- posts 2000 documents with 3 categories each
- users 100 documents

Gatsby output:
```bash
success open and validate gatsby-configs — 0.006 s
success load plugins — 0.271 s
success onPreInit — 0.749 s
success delete html and css files from previous builds — 0.013 s
success initialize cache — 0.014 s
success copy gatsby files — 0.416 s
success onPreBootstrap — 0.032 s
success source and transform nodes — 6.568 s
success building schema — 0.605 s
success createPages — 0.000 s
success createPagesStatefully — 0.086 s
success onPreExtractQueries — 0.002 s
success update schema — 0.212 s
success extract queries from components — 0.040 s
success run graphql queries — 0.051 s — 2/2 40.02 queries/second
success write out page data — 0.032 s
success write out redirect data — 0.004 s
success onPostBootstrap — 0.001 s

info bootstrap finished - 16.121 s

 DONE  Compiled successfully in 4448ms
```

database: five_thousand_items
collections:
- categories 200 documents
- comments 5000 documents
- posts 5000 documents with 3 categories each
- users 500 documents

Gatsby output:

```bash
success open and validate gatsby-configs — 0.014 s
success load plugins — 0.237 s
success onPreInit — 0.659 s
success delete html and css files from previous builds — 0.006 s
success initialize cache — 0.023 s
success copy gatsby files — 0.307 s
success onPreBootstrap — 0.024 s
success source and transform nodes — 8.896 s
success building schema — 0.638 s
success createPages — 0.000 s
success createPagesStatefully — 0.090 s
success onPreExtractQueries — 0.000 s
success update schema — 0.331 s
success extract queries from components — 0.043 s
success run graphql queries — 0.072 s — 2/2 28.37 queries/second
success write out page data — 0.035 s
success write out redirect data — 0.004 s
success onPostBootstrap — 0.001 s

info bootstrap finished - 19.632 s

 DONE  Compiled successfully in 4510ms
```

database: ten_thousand_items
collections:
- categories 500 documents
- comments 10000 documents
- posts 10000 documents with 3 categories each
- users 1000 documents

Gatsby output:
```bash
success open and validate gatsby-configs — 0.014 s
success load plugins — 0.252 s
success onPreInit — 0.555 s
success delete html and css files from previous builds — 0.044 s
success initialize cache — 0.045 s
success copy gatsby files — 0.478 s
success onPreBootstrap — 0.022 s
success source and transform nodes — 16.258 s
success building schema — 0.774 s
success createPages — 0.000 s
success createPagesStatefully — 0.088 s
success onPreExtractQueries — 0.001 s
success update schema — 0.464 s
success extract queries from components — 0.039 s
success run graphql queries — 0.092 s — 2/2 22.36 queries/second
success write out page data — 0.027 s
success write out redirect data — 0.002 s
success onPostBootstrap — 0.003 s

info bootstrap finished - 27.951 s

 DONE  Compiled successfully in 5136ms
```

database: fifty_thousand_items
collections:
- categories 2000 documents
- comments 50000 documents
- posts 50000 documents with 3 categories each
- users 5000 documents

Gatsby output:

```bash
success open and validate gatsby-configs — 0.010 s
success load plugins — 0.237 s
success onPreInit — 0.464 s
success delete html and css files from previous builds — 0.010 s
success initialize cache — 0.019 s
success copy gatsby files — 0.441 s
success onPreBootstrap — 0.024 s
success source and transform nodes — 247.940 s
success building schema — 2.733 s
success createPages — 0.001 s
success createPagesStatefully — 0.073 s
success onPreExtractQueries — 0.000 s
success update schema — 2.190 s
success extract queries from components — 0.038 s
success run graphql queries — 0.247 s — 2/2 8.13 queries/second
success write out page data — 0.030 s
success write out redirect data — 0.001 s
success onPostBootstrap — 0.001 s

info bootstrap finished - 260.558 s

 DONE  Compiled successfully in 6433ms
```

database: one_hundred_thousand_items
collections:
- categories 20000 documents
- comments 100000 documents
- posts 100000 documents with 3 categories each
- users 10000 documents

Gatsby output:
```bash
success open and validate gatsby-configs — 0.008 s
success load plugins — 0.228 s
success onPreInit — 0.364 s
success delete html and css files from previous builds — 0.008 s
success initialize cache — 0.016 s
success copy gatsby files — 0.265 s
success onPreBootstrap — 0.018 s
success source and transform nodes — 815.137 s
success building schema — 7.784 s
success createPages — 0.000 s
success createPagesStatefully — 0.061 s
success onPreExtractQueries — 0.001 s
success update schema — 6.157 s
success extract queries from components — 0.028 s
success run graphql queries — 0.555 s — 2/2 3.62 queries/second
success write out page data — 0.016 s
success write out redirect data — 0.001 s
success onPostBootstrap — 0.001 s

info bootstrap finished - 836.335 s

 DONE  Compiled successfully in 9699ms
```

Working perfectly! As expected the bigger you go and more references you have, the bigger the impact. But that's to be expected. If you or anyone else wants to update the docs with a table or something in that fashion, feel free to take the results of my small test and do it. If additional information is required feel free to say something. Otherwise i'm really hoping to see this merged as soon as possible and with that close the original issue. It goes without saying thank you very much for the time you've invested in this.

**Comment by wesleylhandy** [comment]
@jonniebigodes Awesome work testing this out! I'll add something to documentation and push later today. Are you able to make an approving review or who can we get to approve it at that point?

**Comment by jonniebigodes** [comment]
@wesleylhandy done and done

**Comment by jonniebigodes** [comment]
@wesleylhandy just saw the table with the contents. Looks really awesome.

**Comment by wesleylhandy** [comment]
I'm going to leave this PR alone now. I realized I forgot to set an explicit default for `preserveObjectIds` and I making a copy of each returned document from `cursor.ToArray` I was adding perhaps both space and time complexity to the function. Instead, I am deconstructing each document as `{_id, ...item}` which removes the need to call `delete item._id`.

File: packages/gatsby-source-mongodb/src/__tests__/stringify-object-ids.js [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+    res = [str1, str2, str3],
+    stringifiedArr = stringifyObjectIds(arr)
+
+  expect(stringifiedArr[0]).toEqual(res[0])
+  expect(stringifiedArr[1]).toEqual(res[1])
+  expect(stringifiedArr[2]).toEqual(res[2])
+})
+
+test(`when passed an array of object literals with ObjectIDs it returns an array of object literals with strings`, () => {
+  const str1 = `5c6d85cff2bc0b7a5da61124`,
```
Comment (DSchau) [inline packages/gatsby-source-mongodb/src/__tests__/stringify-object-ids.js]: Same idea here--I think this is a bit manual. If we can re-write it I think that would be preferable, e.g.

```
const ids = [
  // id1
  // id2
  // id3
]

expect(
  stringifyObjectIds(
    ids.map((id, index) => ({
      key: index,
      id: new ObjectID(id)
    })
  )
    .map(obj => obj.id)
).toEqual(ids)
```

File: packages/gatsby-source-mongodb/src/__tests__/stringify-object-ids.js [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+    ],
+    stringifiedObj = stringifyObjectIds(obj)
+
+  expect(stringifiedObj[0].id).toEqual(res[0].id)
+  expect(stringifiedObj[1].id).toEqual(res[1].id)
+  expect(stringifiedObj[2].id).toEqual(res[2].id)
+})
+
+test(`when passed a deeply nested, complex object literal, it returns all stringified ObjectIds`, () => {
+  const str1 = `5c6d85cff2bc0b7a5da61124`,
```
Comment (DSchau) [inline packages/gatsby-source-mongodb/src/__tests__/stringify-object-ids.js]: This one might be a little more challenging to write in a cleaner way--so let me know if you'd like an assist!

File: packages/gatsby-source-mongodb/src/gatsby-node.js [hunk 3]
```diff
@@ -76,11 +78,17 @@ function createNodes(
         reject(err)
       }
 
-      documents.forEach(item => {
-        var id = item._id.toString()
-        delete item._id
+      documents.forEach(({ _id, ...item }) => {
+        const id = _id.toHexString()
```
[the remark this answers is not in the document]
Comment (wesleylhandy) [reply packages/gatsby-source-mongodb/src/gatsby-node.js]: `toString` is not a Node driver function on an ObjectID, http://mongodb.github.io/node-mongodb-native/3.1/api/ObjectID.html

File: packages/gatsby-source-mongodb/README.md [hunk 4]
```diff
@@ -48,7 +48,20 @@ module.exports = {
...
+- **clienOptions**: for setting options on the creation of a `new MongoClient`. By default, to handle the various connection URI's necessary for newer versions of MongoDB Atlas, for instance, we pass { `useNewUrlParser`: `true` }. You can override the default by passing either an empty object literal or filled with other valid connection options These are the subset of connection options that have to be passed as part of an object to the creation of a new `MongoClient` instance: http://mongodb.github.io/node-mongodb-native/3.1/reference/connecting/connection-settings/
+- **preserveObjectIds**: if you use `ObjectID`s to store relationships between documents and collections, set this to the Boolean `true`, ex: preserveObjectIds: `true`. Please note, this uses a recursive algorithm to walk through each document tree and replace instances of `ObjectID` with a valid string. For large datasets, this could slow down the build process.
```
[the remark this answers is not in the document]
Comment (wesleylhandy) [reply packages/gatsby-source-mongodb/README.md]: Good point.

    for preserving nested ObjectIDs within documents, set this to the Boolean `true`, ex: preserveObjectIds: `true`.

Is that too brief?

Should there be any mention of the possibility of a slower build?

File: packages/gatsby-source-mongodb/src/__tests__/stringify-object-ids.js (hunk 2 above)
Comment (wesleylhandy) [reply packages/gatsby-source-mongodb/src/__tests__/stringify-object-ids.js]: I'm not sure how cleaner I got it:

    const ids = [
      new ObjectID().toHexString(),
      new ObjectID().toHexString(),
      new ObjectID().toHexString(),
      new ObjectID().toHexString(),
      new ObjectID().toHexString(),
      new ObjectID().toHexString(),
    ]
    const complexNested = {
      id: new ObjectID(ids[0]),
      connections: [
        {
          id: new ObjectID(ids[1]),
        },
        {
          ids: [new ObjectID(ids[2]), new ObjectID(ids[3])],
        },
      ],
      name: {
        first: `sally`,
        last: {
          maiden: {
            val: `smith`,
            id: new ObjectID(ids[4]),
          },
          current: {
            val: `wilson`,
            id: new ObjectID(ids[5]),
          },
          former: {
            val: `smith`,
            id: new ObjectID(ids[4]),
          },
        },
      },
    }
    const expectedResult = {
      id: ids[0],
      connections: [
        {
          id: ids[1],
        },
        {
          ids: [ids[2], ids[3]],
        },
      ],
      name: {
        first: `sally`,
        last: {
          maiden: {
            val: `smith`,
            id: ids[4],
          },
          current: {
            val: `wilson`,
            id: ids[5],
          },
          former: {
            val: `smith`,
            id: ids[4],
          },
        },
      },
    }

    expect(stringifyObjectIds(complexNested)).toEqual(expectedResult)
    expect(stringifyObjectIds(complexNested)).not.toEqual(complexNested)

**Comment by wesleylhandy** [comment]
@DSchau Hey Dustin, Thanks for the review. I have made several changes as requested to the tests and to the readme. For the tests, I removed all hard coding of magic strings. I used a little more programmatic approach as well. For the readme, I reduced it to once sentence explaining what the feature does. Let me know if you think more info is needed there. Thanks again!

**Comment by DSchau** [comment]
```
Successfully published:
 - gatsby-source-mongodb@2.0.17
```

**Comment by laurenskling** [comment]
Thanks for making this fix @wesleylhandy !
It actually still doesn't really get the relationship in grahpQL though. Only outputs the reference ID as a string (which is more helpful than outputting nothing ofcourse!). It would be great if we could actually lookup the model referenced and allow to query it via GraphQL. That would make mongo relationships in graphQL very powerful! You think this is something we could achieve? When `instanceof ObjectID` hits, why can try to find the actual referenced document and add it's type?

**Comment by jonniebigodes** [comment]
@laurenskling what was implemented is actually how a relationship is implemented in a SQL database and what was proposed by and achieved by @wesleylhandy 's work. With this approach added with this pull request you can adjust and treat it like the Gatsby blog starter. What you intend with your latest comment, could work without any significant issues on smaller datasets. On larger datasets, could lead to issues like heap out of memory and/or extremely long build times or never ending build times. As it would generate a large ammount of duplicate data with each item.

**Comment by wesleylhandy** [comment]
Thanks @jonniebigodes & @DSchau for code review and advice working through this.

**Comment by wesleylhandy** [comment]
@laurenskling It would be interesting but I think instead of rewriting something like `Mongoose` for this plugin, perhaps someone could write a `gatsby-source-mongoose` plugin that takes in `Schema`s and outputs populated `documents`.

**Comment by jonniebigodes** [comment]
@wesleylhandy no need to thank, glad to have helped out.

**Comment by laurenskling** [comment]
> @laurenskling It would be interesting but I think instead of rewriting something like `Mongoose` for this plugin, perhaps someone could write a `gatsby-source-mongoose` plugin that takes in `Schema`s and outputs populated `documents`.

You are absolutely right, that sounds more obvious! Now let's hope a mongoose hero feels the same way 😅 

Thanks for the contributions guys!

**Pull request merged**